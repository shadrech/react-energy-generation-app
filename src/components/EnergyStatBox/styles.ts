import styled, { keyframes } from "styled-components";
import breakpoint from "styled-components-breakpoint";

const fadeIn = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

interface StatBoxProps {
  animationIndex?: number
}

export const StatBox = styled.article<StatBoxProps>`
  --width: 8;
  --columns: 9;
  --gap: calc((var(--columns) - var(--width)) * 1%);
  --boxBasis: calc(var(--width) / var(--columns) * 100%);
  flex-basis: calc(var(--boxBasis) - var(--gap));
  display: flex;
  align-items: center;
  justify-content: space-between;
  background: white;
  margin-bottom: .7em;
  border: 1px solid #e8e9ed;
  padding: 1rem;
  opacity: 0;
  animation: ${fadeIn} .7s ease-out forwards;
  animation-delay: ${({ animationIndex }) => animationIndex ? animationIndex * 0.3 : 0}s;

  ${breakpoint('tablet')`
    --width: 4;
    margin-bottom: calc(var(--gap) / 2);
  `}
  
  ${breakpoint('desktop')`
    --width: 3;
  `}
`;

export const ChartWrapper = styled.div`
  position: relative;
  display: inline-flex;
  align-items: center;
  width: 30%;

  p {
    margin: 0;
    position: absolute;
    left: 50%;
    transform: translateX(-50%);
    font-family: Roboto;
    color: #525454;
    font-size: 1rem;
  }
`;

export const TextWrapper = styled.div`
  position: relative;
  display: inline-flex;
  justify-content: space-around;
  flex-direction: column;
  text-align: right;
  height: 100%;

  h1 {
    margin: 0;
    font-family: 'Courier New', Courier, monospace;
    font-weight: normal;
    font-size: 3rem;
    color: #27CFDB;
  }
  h2 {
    margin: 0;
    font-family: Bookman;
    color: #aaaaaa;
    font-weight: normal;
  }
`;
