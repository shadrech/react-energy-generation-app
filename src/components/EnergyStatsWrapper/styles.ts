import styled from 'styled-components';
import breakpoint from "styled-components-breakpoint";

export const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  width: 100%;
  height: 100%;
  --sidePadding: 3rem;
  --topBottomPadding: 1rem;
  padding: var(--topBottomPadding) var(--sidePadding);
  box-sizing: border-box;
  height: inherit;

  ${breakpoint('tablet')`
    --sidePadding: 5rem;
    --topBottomPadding: 1.5rem;
  `}
  
  ${breakpoint('desktop')`
    --sidePadding: 7rem;
    --topBottomPadding: 2rem;
  `}
`;
