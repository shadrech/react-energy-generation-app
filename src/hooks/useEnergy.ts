import React from 'react'
import axios from 'axios'

type Fuel = 'biomass' | 'coal' | 'imports' | 'gas' | 'nuclear' | 'other' | 'hydro' | 'solar' | 'wind'

export interface GenerationMix {
  fuel: Fuel
  perc: number
}

interface GenerationData {
  from: string
  to: string
  generationmix: GenerationMix[]
}

export interface EnergyData {
  data: GenerationData
  error: Error
  loading: boolean
  fetchData(): Promise<void>
}

export const initialEnergyData: EnergyData = { data: null, error: null, loading: false, fetchData: null };

export default function(): EnergyData {
  const [data, updateData] = React.useState<GenerationData>(null)
  const [error, updateError] = React.useState<Error>(null)
  const [loading, updateLoading] = React.useState<boolean>(false)
  const fetchData = async () => {
    updateLoading(true)
    try {
      const response = await axios.get<Pick<EnergyData, 'data'>>('https://api.carbonintensity.org.uk/generation')
      updateData(response.data.data)
      updateError(null)
    } catch (error) {
      console.error('Error:', error)
      updateError(error)
    }
    updateLoading(false)
  }

  return {
    data,
    error,
    loading,
    fetchData
  }
}
