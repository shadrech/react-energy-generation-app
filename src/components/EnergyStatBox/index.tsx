import React from 'react'
import CountUp from 'react-countup'
import { VictoryPie } from "victory"
import { StatBox, ChartWrapper, TextWrapper } from './styles'
import { GenerationMix } from '../../hooks/useEnergy';

interface Props {
  generation: GenerationMix
  animationIndex?: number
}

const EnergyStatBox: React.FunctionComponent<Props> = ({ generation: { perc, fuel }, animationIndex }) => {
  const animationDelay = animationIndex ? (animationIndex * 0.3) + 1 : 0;
  return (
    <StatBox animationIndex={animationIndex}>
      <ChartWrapper>
        <svg style={{ position: 'absolute' }}>
          <defs>
            <linearGradient id="lgrad" x1="0%" y1="100%" x2="100%" y2="0%" >
            <stop offset="0%" stopColor="rgb(49,214,231)" />
            <stop offset="39%" stopColor="rgb(49,214,231)" />
            <stop offset="100%" stopColor="rgb(188,135,252)" />
            </linearGradient>
          </defs>
        </svg>
        <VictoryPie
          data={[
            { y: perc },
            { y: 100 - perc }
          ]}
          colorScale={[
            'url(#lgrad)',
            '#f4f5f9'
          ]}
          labels={[]}
          innerRadius={180}
        />
        <p>
          <CountUp end={Math.round(perc)} delay={animationDelay} />%
        </p>
      </ChartWrapper>
      <TextWrapper>
          <h1>
            <CountUp end={perc} delay={animationDelay} decimals={1} />
          </h1>
          <h2>{`${fuel[0].toUpperCase()}${fuel.slice(1)}`}</h2>
      </TextWrapper>
    </StatBox>
  )
}

export default EnergyStatBox
