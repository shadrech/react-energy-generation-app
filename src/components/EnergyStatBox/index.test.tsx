import React from 'react'
import { render } from '@testing-library/react'
import wait from 'waait'
import EnergyStatBox from '.'
import { StatBox } from './styles';
import { GenerationMix } from '../../hooks/useEnergy'

const statBoxId = (StatBox as any).styledComponentId

describe('<EnergyStatBox />', () => {

  const generation: GenerationMix = {
    fuel: 'hydro',
    perc: 17
  }

  it('renders correctly as component animates in', async () => {
    const { container } = render(<EnergyStatBox generation={generation} />)
    expect(container).toMatchSnapshot()
    await wait(2000)
    expect(container).toMatchSnapshot()
  })

  it('adds a animation delay if animationIndex prop provided', () => {
    const { container } = render(<EnergyStatBox generation={generation} animationIndex={2} />)
    const statBox = container.getElementsByClassName(statBoxId)[0]
    expect(statBox).toHaveStyle('animation-delay: 0.6s')
  })

  it('does not add a animation delay if animationIndex prop not provided', () => {
    const { container } = render(<EnergyStatBox generation={generation} />)
    const statBox = container.getElementsByClassName(statBoxId)[0]
    expect(statBox).toHaveStyle('animation-delay: 0s')
  })

})
