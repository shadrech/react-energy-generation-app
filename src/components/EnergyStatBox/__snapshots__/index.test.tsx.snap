// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EnergyStatBox /> renders correctly as component animates in 1`] = `
<div>
  <article
    class="sc-bdVaJa kkwavi"
  >
    <div
      class="sc-bwzfXH hPCbED"
    >
      <svg
        style="position: absolute;"
      >
        <defs>
          <lineargradient
            id="lgrad"
            x1="0%"
            x2="100%"
            y1="100%"
            y2="0%"
          >
            <stop
              offset="0%"
              stop-color="rgb(49,214,231)"
            />
            <stop
              offset="39%"
              stop-color="rgb(49,214,231)"
            />
            <stop
              offset="100%"
              stop-color="rgb(188,135,252)"
            />
          </lineargradient>
        </defs>
      </svg>
      <div
        class="VictoryContainer"
        style="height: 100%; width: 100%; pointer-events: none; position: relative;"
      >
        <svg
          height="400"
          role="img"
          style="pointer-events: all; width: 100%; height: auto;"
          viewBox="0 0 400 400"
          width="400"
        >
          <g>
            <path
              d="M1.1021821192326179e-14,-180A180,180,0,0,1,157.73520240789546,-86.71566133830872L131.44600200657956,-72.26305111525727A150,150,0,0,0,9.184850993605149e-15,-150Z"
              role="presentation"
              shape-rendering="auto"
              style="fill: url(#lgrad); padding: 10px; stroke: transparent; stroke-width: 1;"
              transform="translate(200, 200)"
            />
            <path
              d="M157.73520240789546,-86.71566133830872A180,180,0,1,1,1.26806651969044e-13,-180L1.0567220997420333e-13,-150A150,150,0,1,0,131.44600200657956,-72.26305111525727Z"
              role="presentation"
              shape-rendering="auto"
              style="fill: #f4f5f9; padding: 10px; stroke: transparent; stroke-width: 1;"
              transform="translate(200, 200)"
            />
          </g>
        </svg>
        <div
          style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: auto;"
        >
          <svg
            height="400"
            style="overflow: visible; width: 100%; height: auto;"
            viewBox="0 0 400 400"
            width="400"
          />
        </div>
      </div>
      <p>
        <span>
          0
        </span>
        %
      </p>
    </div>
    <div
      class="sc-htpNat iZGEOS"
    >
      <h1>
        <span>
          0.0
        </span>
      </h1>
      <h2>
        Hydro
      </h2>
    </div>
  </article>
</div>
`;

exports[`<EnergyStatBox /> renders correctly as component animates in 2`] = `
<div>
  <article
    class="sc-bdVaJa kkwavi"
  >
    <div
      class="sc-bwzfXH hPCbED"
    >
      <svg
        style="position: absolute;"
      >
        <defs>
          <lineargradient
            id="lgrad"
            x1="0%"
            x2="100%"
            y1="100%"
            y2="0%"
          >
            <stop
              offset="0%"
              stop-color="rgb(49,214,231)"
            />
            <stop
              offset="39%"
              stop-color="rgb(49,214,231)"
            />
            <stop
              offset="100%"
              stop-color="rgb(188,135,252)"
            />
          </lineargradient>
        </defs>
      </svg>
      <div
        class="VictoryContainer"
        style="height: 100%; width: 100%; pointer-events: none; position: relative;"
      >
        <svg
          height="400"
          role="img"
          style="pointer-events: all; width: 100%; height: auto;"
          viewBox="0 0 400 400"
          width="400"
        >
          <g>
            <path
              d="M1.1021821192326179e-14,-180A180,180,0,0,1,157.73520240789546,-86.71566133830872L131.44600200657956,-72.26305111525727A150,150,0,0,0,9.184850993605149e-15,-150Z"
              role="presentation"
              shape-rendering="auto"
              style="fill: url(#lgrad); padding: 10px; stroke: transparent; stroke-width: 1;"
              transform="translate(200, 200)"
            />
            <path
              d="M157.73520240789546,-86.71566133830872A180,180,0,1,1,1.26806651969044e-13,-180L1.0567220997420333e-13,-150A150,150,0,1,0,131.44600200657956,-72.26305111525727Z"
              role="presentation"
              shape-rendering="auto"
              style="fill: #f4f5f9; padding: 10px; stroke: transparent; stroke-width: 1;"
              transform="translate(200, 200)"
            />
          </g>
        </svg>
        <div
          style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: auto;"
        >
          <svg
            height="400"
            style="overflow: visible; width: 100%; height: auto;"
            viewBox="0 0 400 400"
            width="400"
          />
        </div>
      </div>
      <p>
        <span>
          17
        </span>
        %
      </p>
    </div>
    <div
      class="sc-htpNat iZGEOS"
    >
      <h1>
        <span>
          17.0
        </span>
      </h1>
      <h2>
        Hydro
      </h2>
    </div>
  </article>
</div>
`;
